Sources/NIOHTTP1/NIOHTTPClientUpgradeHandler.swift

```diff
@@ -0,0 +1,404 @@
+    
+    private var upgradeState: UpgradeState = .requestRequired
+    
+    private var receivedMessages: CircularBuffer<NIOAny> = CircularBuffer()
+    
+    /// Create a `HTTPClientUpgradeHandler`.
+    ///
+    /// - Parameter upgraders: All `HTTPClientProtocolUpgrader` objects that will add their upgrade request
+    ///     headers and handle the upgrade if there is a response for their protocol. They should be placed in
+    ///     order of the preference for the upgrade.
+    /// - Parameter httpHandlers: All `RemovableChannelHandler` objects which will be removed from the pipeline
+    ///     once the upgrade response is sent. This is used to ensure that the pipeline will be in a clean state
+    ///     after the upgrade. It should include any handlers that are directly related to handling HTTP.
+    ///     At the very least this should include the `HTTPEncoder` and `HTTPDecoder`, but should also include
+    ///     any other handler that cannot tolerate receiving non-HTTP data.
+    /// - Parameter upgradeCompletionHandler: A closure that will be fired when HTTP upgrade is complete.
+    public init(upgraders: [NIOHTTPClientProtocolUpgrader],
+                httpHandlers: [RemovableChannelHandler],
+                upgradeCompletionHandler: @escaping (ChannelHandlerContext) -> Void) {
+
+        precondition(upgraders.count > 0, "A minimum of one protocol upgrader must be specified.")
+
+        self.upgraders = upgraders
+        self.httpHandlers = httpHandlers
+        self.upgradeCompletionHandler = upgradeCompletionHandler
+    }
+
+    public func write(context: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
+
+        switch self.upgradeState {
+        
+        case .requestRequired:
+            let updatedData = self.addHeadersToOutboundOut(data: data)
+            context.write(updatedData, promise: promise)
+
+        case .awaitingConfirmationResponse:
+            // Still have full http stack.
+            context.write(data, promise: promise)
+            
+        case .upgraderReady, .upgrading:
+            /// We shouldn't be sending anything else until we get a response.
+            context.fireErrorCaught(NIOHTTPClientUpgradeError.writingToHandlerDuringUpgrade)
+
+        case .upgradingAddingHandlers:
+            // These are most likely messages immediately fired by a new protocol handler.
+            // As that is added last we can just forward them on.
+            context.write(data, promise: promise)
+            
+        case .upgradeComplete:
+            //Upgrade complete and this handler should have been removed from the pipeline.
+            context.fireErrorCaught(NIOHTTPClientUpgradeError.writingToHandlerAfterUpgradeCompleted)
+        case .upgradeFailed:
+            //Upgrade failed and this handler should have been removed from the pipeline.
+            context.fireErrorCaught(NIOHTTPClientUpgradeError.writingToHandlerAfterUpgradeCompleted)
```

Comment:
These error states need to be tested to confirm that they fail the write promises as well as firing errors through the pipeline. Right now the write promises get dropped, and that's a fatal error in a NIO program.